src/components/Tooltip/Tooltip.tsx

```diff
@@ -0,0 +1,369 @@
+import React, {
+  createElement,
+  ForwardRefExoticComponent,
+  ReactElement,
+  ReactNode,
+  RefObject,
+  useEffect,
+  useRef,
+  useState,
+} from 'react'
+import classnames from 'classnames'
+import { isElementInViewport } from './utils'
+
+type TooltipProps<T> = {
+  label: string
+  position?: 'top' | 'bottom' | 'left' | 'right' | undefined
+  className?: string
+  children: ReactNode
+} & T
+
+type TooltipTriggerProps<T> = {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  ref: RefObject<any>
+  onMouseEnter: (e: React.MouseEvent) => void
+  onMouseOver: (e: React.MouseEvent) => void
+  onFocus: (e: React.FocusEvent) => void
+  onMouseLeave: (e: React.MouseEvent) => void
+  onBlur: (e: React.FocusEvent) => void
+  onKeyDown: (e: React.KeyboardEvent) => void
+  children: ReactNode
+} & T
+
+interface WithCustomTooltipProps {
+  asCustom: // eslint-disable-next-line @typescript-eslint/no-explicit-any
```

Comment:
They're not 100% the same but I used [this Formik source code](https://github.com/formium/formik/blob/08d41a7d78290c272ac484a29828f54b74ec4870/packages/formik/src/Field.tsx#L33) as an example of typing a polymorphic component prop.

Other references that have been helpful in figuring this out:
- https://blog.andrewbran.ch/polymorphic-react-components/
- https://react-typescript-cheatsheet.netlify.app/docs/advanced/patterns_by_usecase
- https://github.com/kripod/react-polymorphic-box
- https://github.com/typescript-cheatsheets/react/issues/167#issuecomment-751347673